dm/master/scheduler/scheduler_test.go

```diff
@@ -719,3 +719,147 @@ func (t *testScheduler) TestRestartScheduler(c *C) {
+	sourceCfg2.SourceID = sourceID2
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// step 1: start an empty scheduler without listening the worker event
+	s.started = true
+	s.cancel = cancel
+	s.etcdCli = etcdTestCli
+
+	// step 2: add two sources and register four workers
+	c.Assert(s.AddSourceCfg(sourceCfg1), IsNil)
+	c.Assert(s.AddSourceCfg(sourceCfg2), IsNil)
+	c.Assert(s.unbounds, HasLen, 2)
+	c.Assert(s.unbounds, HasKey, sourceID1)
+	c.Assert(s.unbounds, HasKey, sourceID2)
+
+	c.Assert(s.AddWorker(workerName1, workerAddr1), IsNil)
+	c.Assert(s.AddWorker(workerName2, workerAddr2), IsNil)
+	c.Assert(s.AddWorker(workerName3, workerAddr3), IsNil)
+	c.Assert(s.AddWorker(workerName4, workerAddr4), IsNil)
+	c.Assert(s.workers, HasLen, 4)
+	c.Assert(s.workers, HasKey, workerName1)
+	c.Assert(s.workers, HasKey, workerName2)
+	c.Assert(s.workers, HasKey, workerName3)
+	c.Assert(s.workers, HasKey, workerName4)
+
+	// step 3: add two workers, and then cancel them to simulate they have lost connection
+	var wg sync.WaitGroup
+	ctx1, cancel1 := context.WithCancel(ctx)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		c.Assert(ha.KeepAlive(ctx1, etcdTestCli, workerName1, keepAliveTTL), IsNil)
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		c.Assert(ha.KeepAlive(ctx1, etcdTestCli, workerName2, keepAliveTTL), IsNil)
+	}()
+	c.Assert(utils.WaitSomething(30, 10*time.Millisecond, func() bool {
+		kam, _, err := ha.GetKeepAliveWorkers(etcdTestCli)
+		return err == nil && len(kam) == 2
+	}), IsTrue)
+	cancel1()
+	wg.Wait()
+	// check whether keepalive lease is out of date
+	time.Sleep(time.Duration(keepAliveTTL) * time.Second)
+	var rev int64
+	c.Assert(utils.WaitSomething(30, 10*time.Millisecond, func() bool {
+		kam, rev1, err := ha.GetKeepAliveWorkers(etcdTestCli)
+		rev = rev1
+		return err == nil && len(kam) == 0
+	}), IsTrue)
+
+	// step 4: trigger etcd compaction and check whether we can receive it through watcher
+	var startRev int64 = 1
+	_, err := etcdTestCli.Compact(ctx, rev)
+	c.Assert(err, IsNil)
+	workerEvCh := make(chan ha.WorkerEvent, 10)
+	workerErrCh := make(chan error, 10)
+	ha.WatchWorkerEvent(ctx, etcdTestCli, startRev, workerEvCh, workerErrCh)
+	select {
+	case err := <-workerErrCh:
+		c.Assert(err, Equals, etcdErrCompacted)
+	case <-time.After(300 * time.Millisecond):
+		c.Fatal("fail to get etcd error compacted")
+	}
+
+	// step 5: scheduler start to handle workerEvent from compact revision, should handle worker keepalive events correctly
+	ctx2, cancel2 := context.WithCancel(ctx)
+	// step 5.1: start one worker before scheduler start to handle workerEvent
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		c.Assert(ha.KeepAlive(ctx2, etcdTestCli, workerName3, keepAliveTTL), IsNil)
+	}()
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		kam, _, err := ha.GetKeepAliveWorkers(etcdTestCli)
+		if err != nil {
+			if _, ok := kam[workerName3]; ok {
+				return len(kam) == 1
+			}
```

Comment:
🤔 checking `kam` when `err != nil`?

Reply:
My mistask... revised in https://github.com/pingcap/dm/pull/499/commits/da66dcf78a2e476798a53c014caa8c1447133dfc